index.bs

```diff
@@ -1799,17 +1799,20 @@ Follow the <a href="https://www.w3.org/2001/tag/doc/promises-guide#one-time-even
 in the <strong><a href="https://www.w3.org/2001/tag/doc/promises-guide">Writing
 Promise-Using Specifications</a></strong> guideline.
 
-<h3 id="promises-and-events">Events should fire before Promises resolve</h3>
+<h3 id="promises-and-events">Events should fire before related Promises resolve</h3>
 
 If a Promise-based asynchronous algorithm dispatches events,
 it should dispatch them before the Promise resolves,
 rather than after.
 
-This guarantees that once the Promise resolves,
-all effects of the algorithm have been applied.
-For example, if an author changes some state
-in reaction to an event which the Promise dispatches,
-they can be sure that all of the state is consistent if the Promise is resolved.
+When a promise is resolved, a microtask is queued to run its reaction callbacks.
+Microtasks are processed when the JavaScript stack empties.
+Dispatching an event can involve the JavaScript stack emptying between listeners.
+As a result, if a promise is resolved before dispatching a related event,
+the promise reaction callbacks will be invoked between the first and second listeners of the event.
```

Comment:
Hey @jakearchibald, sorry to perform necromancy on this one.  We think that it's a distinct improvement, but have some requests.

Is there a reference you could use for this, or an example?  This took me several reads to understand properly.  It's accurate, but the effect is surprising.

The gap, I think, is that while you explain how promises can end up interleaved with event handlers (is it a handler or a listener?) there is a gap regarding the execution model.  That doesn't need to be here, but a reference would help.

(Questions I couldn't answer is what the effect on bubbling events.  And maybe what model event listener dispatch takes.)

Reply:
I'm not sure what's missing, or what reference I'd use. Let me try and explain it in more detail, and you can tell me which bit(s), if any, give you an "aha!" moment, and I'll work that into the description.

Pending promise reaction callbacks are processed in microtasks. There are many places that microtasks are processed, but it's always when the JS stack is empty. The most-encountered place that microtasks are processed is https://html.spec.whatwg.org/multipage/webappapis.html#clean-up-after-running-script, which is called whenever the JavaScript stack empties. This algorithm gets called whenever script or callbacks are executed, eg https://webidl.spec.whatwg.org/#invoke-a-callback-function.

Imagine the following spec:

1. Resolve promise.
2. Dispatch event.
3. Log "hello".

Resolving a promise is async, so the promise callbacks aren't called as part of step 1.

However, dispatching an event is synchronous. _All_ the related callbacks for the event (listeners, handlers, capturing, bubbling) are called before we get to step 3.

Let's say there are two such callbacks to call as part of step 2. Each listener/handler callback is [invoked](https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke), which calls [call a user object's operation](https://webidl.spec.whatwg.org/#call-a-user-objects-operation), which calls [clean up after running script](https://html.spec.whatwg.org/multipage/webappapis.html#clean-up-after-running-script) when each callback is complete.

Because [clean up after running script](https://html.spec.whatwg.org/multipage/webappapis.html#clean-up-after-running-script) causes microtasks to be processed, the order callbacks are called is:

1. The first event listener callback
2. All currently queued reaction callbacks for the promise
3. The second event listener callback

That seems bad/confusing.

If the spec was:

1. Dispatch event.
2. Resolve promise.
3. Log "hello".

Then all event callbacks will have been called before we get to step 2. That means the order of callbacks would be:

1. The first event listener callback
2. The second event listener callback
3. …potentially some other stuff…
4. All currently queued reaction callbacks for the promise

I've added "potentially some other stuff", since it's possible for other things to happen before the next microtask checkpoint. There's a microtask checkpoint at the end of every task. There are checkpoints in the rendering steps too. And again, whenever the JS stack empties.

Reply:
Thanks for indulging me @jakearchibald. That is very helpful.

> dispatching an event is synchronous

This is the missing piece.  (To whoever thought this was a good idea: I hope you understand why this makes me very sad and feel at least a little remorse.  Of course I forgive you, because I understand that you don't always know better when you ship something.)

My expectation was that - if we can use threading terms - you would signal the main JS thread when you want to fire an event and it would be enqueued for processing.  That it is synchronous is a little horrifying, frankly, but I can see how it ended up that way.

Reply:
@martinthomson To be clear, end user input is definitely scheduled in that way (see the event loop and task concepts in the HTML Standard). Event dispatch is a small "synchronous" part of that (and is defined in the DOM Standard).

Reply:
Yeah, I think we're running into sync vs async being kinda fluffy terms when you actually dig into them.

Some events are supposed to fire at the 'same time' because they're related to the same thing, like `input` and `change` on some inputs. This is only possible because dispatching events is synchronous. Although a task could be queued to make those calls, making it feel async depending on where you're standing.

`event.dispatchEvent` is definitely synchronous though, which is how you can tell if the event was cancelled in the next line.